README.md
web2project-mantis
==================

(This content is lifted from the web2project forums and is provided here without comment or correction for background purposes.)

Source: http://forums.web2project.net/viewtopic.php?p=4808#p4808

Web2Project Mantis Addon

Module Overview
This module provides a way for a user to raise issues within Mantis in relation to a W2P projec

How does it work ?
Within W2P, you will find a Tab called Mantis when looking at a given project.
For each W2P project, a Mantis project will be created (if it does not yet exist) when the first issue is added.

Creation of Projects in mantis can be automatic or by using a custom field within Web2Project
In case defined as automatic, Projects in Mantis will be created with a special name :
1. Prefix as defined in config.php
2. W2P-project-id
3. dot
4. W2P-task-id
5. Space
6. W2P-project-name
In case defined as Custom, Projects in Mantis will be created as follows :
Bsed upon a Custom field such that the name can begiven freely:
1. First of all, a one time action, a custom field needs to be created in W2P on project level.
2. This field needs to be called "Mantis" and type should be TEXT.
Now you have an additional field in your project definition where you can define the projectname for Mantis.
Based on vanilla W2P
1. Project name in mantis will be identical to the projectname in W2P.
Be aware that the setting is NOT case sensitive.

If the system is configured to "Custom", there is no use in activating mantis in the top menu bar, it will generate no results.
In that case, only results can be expected when selecting the corresponding project tab.

On the tab all issues related to the project will be shown with the following information:
1. Issue-id in Mantis
2. Date-Time reported
3. Reporter of issue
4. Status
5. Summary
6. Description
The summary actually is a hyperlink which takes you straight into Mantis to maintain the issue.

There is also a button which allows you to add an issue to the project. All you need to fill in is the Summary and the description. Upon saving the issue is stored in Mantis.
The current W2P-user is checked by username in Mantis. If this person does not exist, a user record is created in Mantis.
An email for direct access to Mantis with the password will be send to this user (if configured).
The current project is checked and if it does not yet exists, it will be created.
For the issue the following items are recorded :
1. Project-id
2. Reporter-id
3. Summary
4. Description
5. Date-submitted
6. Date-last-updated
All other activity is left within Mantis.

This module is available for download here: http://web2project.net/forums/viewtopic.php?t=1843

Release Notes
V0.9 30-03-2010
Initial release, converted from DP.
In addition this version is only suitable for Mantis version 1.2.0 and above.

Changelog.txt
Version 2.1.0
This version is geared towards version 2.1 of W2P
Reason for update is different handling of contacts in this version.
Version explanation:
X.Y.Z
X = main version W2P
Y = Second level W2P
Z = Version of module

Version 2.2.1
added time in the mantis datefield when showing list of issues

Version 2.2.3
added option to autocreate tasks
added option to transform Issue into Task

Version 2.3.1 (future)
move config into database opposite manual configuration
make fully use of mantis routines for adding and handling a new issue (including mail notifications)

index.php
<?php
if (!defined('W2P_BASE_DIR')) {
	die('You should not access this file directly');
}
##
## Mantis integration
## (c) Copyright 2010
## Cas Nuy
## www.nuy.info
##

$AppUI->savePlace();
$titleBlock = new CTitleBlock( 'Mantis', 'mantis_logo_button.gif', $m, "$m.$a" );
$titleBlock->show();

$project_id = (int) w2PgetParam($_REQUEST, 'project_id', 0);

$user = new CUser();
$user->load($AppUI->user_id);
$username = $user->user_username;

$query2= "SELECT method_value FROM contacts_methods WHERE method_name='email_primary' and contact_id = " . $AppUI->user_id;
$result2 = mysql_query( $query2 )or die(mysql_error());
while ($row2 = mysql_fetch_array($result2, MYSQL_NUM)) {
	$email= $row2[0];
}

?>
<table width="100%" cellspacing="1" cellpadding="0" border="0">
<tr>
	<td align="right"><input class="button" type="button" value="<?php echo $AppUI->_('Add Issue');?>" onclick="window.location='?m=mantis&a=addissue&projectid=<?php echo $project_id?>'"></td>
</table>


<?php

function show_issue($summary,$bugid,$username,$email)
{
    $mantispath =  w2PgetConfig( 'mantis_path') ;
    $parm = $username ;
    $parm .="||";
    $parm .= $bugid;
    $parm .="||";
    $parm .= $email;
    $parm .="||";
    $parm = base64_encode($parm) ;
    $link  = "<a href=" ;
    $link .= $mantispath ;
    $link .= "/index_dp.php?parm=";
    $link .= $parm ;
    $link .= " target=_blank";
    $link .= ">";
    $link .= $summary ;
    $link .= "</a>";
    return $link;
}

function transform_issue($summary,$desc,$bugid)
{
    global $project_id;

    $link  = "<a href=" ;
    $link .= "index.php?m=mantis&a=addTask&projectid=";
    $link .= $project_id ;
    $link .= "&bugid=";
    $link .= $bugid ;
    $link .= ">";
    $link .= $bugid ;
    $link .= "</a>";
    return $link;
}

// get the definitions
$prefix =  w2PgetConfig( 'mantis_w2p_pref') ;
$mantisprefix =  w2PgetConfig( 'mantis_prefix') ;

$mantislink =  w2PgetConfig( 'mantis_link') ;
$mantistime =  w2PgetConfig( 'mantis_time') ;

$mantis_bug_table = $mantisprefix ;
$mantis_bug_table .= '_bug_table'  ;

$mantis_bug_text_table = $mantisprefix ;
$mantis_bug_text_table .= '_bug_text_table'  ;

$mantis_user_table = $mantisprefix ;
$mantis_user_table .= '_user_table'  ;

$mantis_project_table = $mantisprefix ;
$mantis_project_table .= '_project_table'  ;

// next get the mantis project id based upon the name of project or task
if ($mantislink=="A"){
	$proj =$prefix ;
	$proj .= $project_id ;
	$proj .= '%'  ;
}else{
	$query3="select value_charvalue from custom_fields_values,custom_fields_struct where value_object_id=$project_id and value_field_id=field_id and field_name='Mantis' ";
	$result3 = mysql_query( $query3 )or die(mysql_error());
	while ($row3 = mysql_fetch_array($result3, MYSQL_NUM)) {
		$proj = $row3[0] ;
	}
	if (!$proj){
        $project = new CProject();
        $project->load($project_id);
        $projname = $project->project_name;
        $proj = $project->project_name;
	}
}

$proj1 =$prefix ;
$proj1 .= '%'  ;

// first connect to the correct database
db_connect( $w2Pconfig['mantis_dbhost'], $w2Pconfig['mantis_dbname'],
	$w2Pconfig['mantis_dbuser'], $w2Pconfig['mantis_dbpass'], $w2Pconfig['dbpersist'] );


$idprj = 0;
$query1= "SELECT id FROM $mantis_project_table WHERE name LIKE '$proj' " ;
$result1 = mysql_query( $query1 )or die(mysql_error());
if ($result1){
	while ($row1 = mysql_fetch_array($result1, MYSQL_NUM)) {
		$idprj = $row1[0];
	}
}
// retrieve the issue header/detail/user
if ($idprj !=0 or $project_id>0){
	$query2  = "SELECT $mantis_bug_table.id,summary,description,realname,date_submitted,status " ;
	$query2 .= "FROM $mantis_bug_table,$mantis_bug_text_table,$mantis_user_table ";
	$query2 .= "WHERE $mantis_bug_table.project_id= $idprj  and " ;
	$query2 .= "$mantis_bug_table.id= $mantis_bug_text_table.id and ";
	$query2 .= "reporter_id= $mantis_user_table.id ";
}else{
	// when using custom fields ensure to generate no results
	if ($mantislink <> "A"){
		$proj1 = "no link possible ";
	}

	$query2  = "SELECT $mantis_bug_table.id,summary,$mantis_bug_text_table.description,realname,date_submitted,$mantis_bug_table.status " ;
	$query2 .= "FROM $mantis_bug_table,$mantis_bug_text_table,$mantis_user_table, $mantis_project_table ";
	$query2 .= "WHERE $mantis_bug_table.project_id=$mantis_project_table.id and $mantis_project_table.name like '$proj1'  and " ;
	$query2 .= "$mantis_bug_table.id= $mantis_bug_text_table.id and ";
	$query2 .= "reporter_id= $mantis_user_table.id order by date_submitted";
}

$result2 = mysql_query( $query2 )or die(mysql_error());
?>
<table width="100%" border="0" cellpadding="2" cellspacing="1" class="tbl">
<tr>
	<th width="50"><?php echo $AppUI->_('ID');?></th>
	<th width="50"><?php echo $AppUI->_('Date');?></th>
	<th nowrap="nowrap"><?php echo $AppUI->_('User');?></th>
	<th nowrap="nowrap"><?php echo $AppUI->_('Status');?></th>
	<th nowrap="nowrap"><?php echo $AppUI->_('Summary');?></th>
	<th nowrap="nowrap"><?php echo $AppUI->_('Description');?>&nbsp;&nbsp;</th>
</tr>
<?php
if ($result2){
while ($row2 = mysql_fetch_array($result2, MYSQL_NUM)) {
	$bugid= trim($row2[0]);
	$bugsummary= trim($row2[1]);
	$bugdesc= trim($row2[2]);
	$buguser= trim($row2[3]);
	$bugdate= date("d-M-Y H:i:s" ,trim($row2[4]));
	$bugstat= trim($row2[5]);

	switch ($bugstat) {
	case 10:
		$status = 'New';
		$status_color = '#ffa0a0'; # red
		break;
	case 20:
		$status = 'Feedback';
		$status_color = '#ff50a8'; # purple
		break;
	case 30:
		$status = 'Acknowledged';
		$status_color = '#ffd850'; # orange
		break;
	case 40:
		$status = 'Confirmed';
		$status_color =  '#ffffb0'; # yellow
		break;
	case 50:
		$status = 'Assigned';
		$status_color = '#c8c8ff'; # blue
		break;
	case 80:
		$status = 'Resolved';
		$status_color = '#cceedd'; # buish-green
		break;
	case 90:
		$status = 'Closed';
		$status_color = '#e8e8e8'; # light gray
		break;
	}

?>

<tr>
	<td><?php echo transform_issue($bugsummary,$bugdesc,$bugid) ?></td>
	<td><?php echo $bugdate ?></td>
	<td><?php echo $buguser ?></td>
	<td><strong><b><font color=<?php echo "$status_color";?>>
	<?php echo $status ?>
	</b></strong></td>
	<td><?php echo show_issue($bugsummary,$bugid,$username,$email) ?></td>
	<td><?php echo $bugdesc ?></td>
</tr>
<?php
	}
}
// connect again to DP  database
db_connect( $w2Pconfig['dbhost'], $w2Pconfig['dbname'],
	$w2Pconfig['dbuser'], $w2Pconfig['dbpass'], $w2Pconfig['dbpersist'] );

?>
</table>

addTask.php
<?php
if (!defined('W2P_BASE_DIR')) {
	die('You should not access this file directly');
}

$bug_id = (int) w2PgetParam($_REQUEST, 'bugid', 0);
$project_id = (int) w2PgetParam($_REQUEST, 'projectid', 0);

// get the data from mantis
db_connect( $w2Pconfig['mantis_dbhost'], $w2Pconfig['mantis_dbname'],$w2Pconfig['mantis_dbuser'], $w2Pconfig['mantis_dbpass'], $w2Pconfig['dbpersist'] );
// get the definitions
$prefix =  w2PgetConfig( 'mantis_w2p_pref') ;
$mantisprefix =  w2PgetConfig( 'mantis_prefix') ;
$bug_table = $mantisprefix ;
$bug_table .= '_bug_table'  ;
$text_table = $mantisprefix ;
$text_table .= '_bug_text_table'  ;

$sql = "select summary,description,additional_information from $bug_table,$text_table where $bug_table.bug_text_id=$text_table.id and $bug_table.id =";
$sql .= $bug_id;

$oktask = db_exec($sql) ;
$row1=db_fetch_row($oktask);
$summary		= $row1['summary'];
$description	= $row1['description'];
$description	.= " ";
$description	.= $row1['additional_information'];

// connect again to W2P  database
db_connect( $w2Pconfig['dbhost'], $w2Pconfig['dbname'],$w2Pconfig['dbuser'], $w2Pconfig['dbpass'], $w2Pconfig['dbpersist'] );

$summary .= " ( $bug_id )";

// now create the task
$task = new CTask();
$task->task_name = $summary;
$task->task_project = $project_id;
$task->task_description = $description;
$task->task_owner = $AppUI->user_id;
$task->store();

$link = "m=tasks&a=addedit&task_id=" . $task->task_id;
$AppUI->redirect("$link");

index_dp.php
<?php

$reqVar = '_' . $_SERVER['REQUEST_METHOD'];
$form_vars = $$reqVar;
$parm = $form_vars['parm'] ;
$parm= base64_decode($parm) ;
$url1 = explode("||", $parm);

$f_username = $url1[0];
$f_id = $url1[1];
$email= $url1[2];

require_once( 'core.php' );

$t_user_table = db_get_table( 'mantis_user_table' );
$f_perm_login='false';

$query = "SELECT  password FROM $t_user_table WHERE username='$f_username'";
$result = db_query( $query );
$f_password = db_result( $result );

if ( auth_attempt_login( $f_username, $f_password, $f_perm_login ) ) {
	if ($f_id ==0){
		print_header_redirect( 'main_page.php' );
	} else {
		print_header_redirect( 'view.php?id='.$f_id.'' );
	}
	$t_redirect_url = 'login_cookie_test.php?return=' . $f_return;

}

$hack_pwd = ranpass() ;
if (user_create ( $f_username,"$hack_pwd", "$email", null,false,true,$f_username )) {
	if ( auth_attempt_login( $f_username, "$hack_pwd" , $f_perm_login ) ) {
		// update table with e-mail address when created an account
		$cookie= ranpass(64);
		$query = "Update $t_user_table set email='$email',cookie_string='$cookie' WHERE username='$f_username'";
		$result = db_query( $query );
		if ($f_id == 0 ) {
			print_header_redirect( 'main_page.php' );
		}else {
			print_header_redirect( 'view.php?id='.$f_id.'' );

		}
		$t_redirect_url = 'login_cookie_test.php?return=' . $f_return;
	}
}


function ranpass($len = "8"){
    $pass = NULL;
    for ($i=0; $i<$len; $i++) {
        $char = chr(rand(48,122));
        while (!ereg("[a-z0-9]", $char)) {
            if($char == $lchar) {
                continue;
            }
            $char = chr(rand(48,90));
        }
        $pass .= $char;
        $lchar = $char;
    }
    return $pass;
}

addissue.php
<?php
if (!defined('W2P_BASE_DIR')) {
	die('You should not access this file directly');
}

$action = w2PgetParam($_REQUEST, 'action', '');
$project_id = (int) w2PgetParam($_REQUEST, 'projectid', 0);

if($action) {
	$issue_summary = w2PgetParam($_POST, 'issue_summary', '');
	$issue_description = w2PgetParam($_POST, 'issue_description', '');
    $project_id = (int) w2PgetParam($_REQUEST, 'issue_project', 0);
	$create_task = w2PgetParam($_POST, 'create_task', '');

	$link = "NONE" ;
	if( $action == 'add' ) {
		// first retrieve username
        $user = new CUser();
        $user->load($AppUI->user_id);
        $username = $user->user_username;
        
		$query1= "SELECT method_value FROM contacts_methods WHERE method_name='email_primary' and contact_id = " . $AppUI->user_id;
		$result1 = mysql_query( $query1 )or die(mysql_error());
		while ($row1 = mysql_fetch_array($result1, MYSQL_NUM)) {
			$email= $row1[0];
		}
		// next retrieve projectname
        $project = new CProject();
        $project->load($project_id);
        $projname = $project->project_name;

		// first connect to the correct database
		db_connect( $w2Pconfig['mantis_dbhost'], $w2Pconfig['mantis_dbname'],$w2Pconfig['mantis_dbuser'], $w2Pconfig['mantis_dbpass'], $w2Pconfig['dbpersist'] );
		// get the definitions
		$prefix =  w2PgetConfig( 'mantis_w2p_pref') ;
		$mantisprefix =  w2PgetConfig( 'mantis_prefix') ;
		$mantis_bug_table = $mantisprefix ;
		$mantis_bug_table .= '_bug_table'  ;
		$mantis_bug_text_table = $mantisprefix ;
		$mantis_bug_text_table .= '_bug_text_table'  ;
		$mantis_user_table = $mantisprefix ;
		$mantis_user_table .= '_user_table'  ;
		$mantis_project_table = $mantisprefix ;
		$mantis_project_table .= '_project_table'  ;
		$mantislink =  w2PgetConfig( 'mantis_link') ;
		// next get the mantis project id based upon the name of project or task
		if ($mantislink == "A"){
			$proj  = $prefix ;
			$proj .= $project_id ;
			$proj .= ' '  ;
			$proj .= $projname ;
		} else {
			// connect to the DP database
			db_connect( $w2Pconfig['dbhost'], $w2Pconfig['dbname'],$w2Pconfig['dbuser'], $w2Pconfig['dbpass'], $w2Pconfig['dbpersist'] );
			$query3="select value_charvalue from custom_fields_values,custom_fields_struct where value_object_id=$project_id and value_field_id=field_id and field_name='Mantis' ";
			$result3 = mysql_query( $query3 )or die(mysql_error());
			//  connect to the mantis database
			db_connect( $w2Pconfig['mantis_dbhost'], $w2Pconfig['mantis_dbname'],$w2Pconfig['mantis_dbuser'], $w2Pconfig['mantis_dbpass'], $w2Pconfig['dbpersist'] );
			while ($row3 = mysql_fetch_array($result3, MYSQL_NUM)) {
				$proj = $row3[0] ;
			}
			if (!$proj){
				$proj = $projname;
			}
		}
		// verify if the project exists, if not add
		$query1= "SELECT id FROM $mantis_project_table WHERE name like '$proj' " ;
		$idprj=0;
		$result1 = mysql_query( $query1 )or die(mysql_error());
		while ($row1 = mysql_fetch_array($result1, MYSQL_NUM)) {
			$idprj = $row1[0];
		}
		if ($idprj==0){
			$add1 = "insert into $mantis_project_table (name) values('$proj')";
			$adding1 = mysql_query( $add1 )or die(mysql_error());
			$idprj = mysql_insert_id();
		}
		// verify if the user exists, if not add
		$query2= "SELECT id FROM $mantis_user_table WHERE username = '$username' " ;
		$iduser=0;
		$result2 = mysql_query( $query2 )or die(mysql_error());
		while ($row2 = mysql_fetch_array($result2, MYSQL_NUM)) {
			$iduser = $row2[0];
		}
		if ($iduser==0){
			$pwd=mantispass();
			$pwd1=$pwd;
			if (w2PgetConfig( 'mantis_login')==MD5) {
				$pwd1=md5($pwd);
			}
			$add2 = "insert into $mantis_user_table (username, password,email) values('$username','$pwd','$email')";
			$adding2 = mysql_query( $add2 )or die(mysql_error());
			$iduser = mysql_insert_id();
			if ( w2PgetConfig( 'mantis_mail')==ON) {
				$tekst="Your password for this Mantis site  ";
				$tekst .= " is : ";
				$tekst .= $pwd ;
				$subject = "Password for the Mantis site, user ";
				$subject .= $username ;
				mail($email, $subject, $tekst);
			}
		}

		// insert the issue
		$basenow	= time(true);
		$add3 = "insert into $mantis_bug_table (project_id, reporter_id, summary,date_submitted,last_updated,due_date) values('$idprj','$iduser','$issue_summary',$basenow,$basenow,$basenow)";
		$adding3 = mysql_query( $add3 )or die(mysql_error());
		$idbug = mysql_insert_id();

		// update the issue
		$updat1 = "update $mantis_bug_table set bug_text_id='$idbug' where id='$idbug'";
		$updating1 = mysql_query( $updat1 )or die(mysql_error());

		// insert the text
		$add4 = "insert into $mantis_bug_text_table (id, description) values('$idbug','$issue_description')";
		$adding4 = mysql_query( $add4 )or die(mysql_error());

		// connect again to W2P  database
		db_connect( $w2Pconfig['dbhost'], $w2Pconfig['dbname'],$w2Pconfig['dbuser'], $w2Pconfig['dbpass'], $w2Pconfig['dbpersist'] );
		
		// now create the task if required
		if ( $create_task == 1) {
            $task = new CTask();
            $task->task_name = $issue_summary;
            $task->task_project = $project_id;
            $task->task_description = $issue_description;
            $task->task_owner = $AppUI->user_id;
            $task->store();

            $link = "m=tasks&a=addedit&task_id=" . $task->task_id;
		}
	}
	if ($link == "NONE"){
		if ($project_id==0){
			$link = "m=projects&a=view" ;
		}else{
			$link= "m=projects&a=view&project_id=" . $project_id ;
		}
	}
	$AppUI->redirect("$link");
}

function mantispass($len = "6") {
    $pass = NULL;
    for ($i=0; $i<$len; $i++) {
        $char = chr(rand(48,122));
        while (!ereg("[a-zA-Z0-9]", $char)){
            if ($char == $lchar) {
                continue;
            }
            $char = chr(rand(48,90));
        }
        $pass .= $char;
        $lchar = $char;
    }
    return $pass;
}

?>

<form name="AddEdit" method="post">
    <input name="action" type="hidden" value="add" >
    <input name="issue_project" type="hidden" value=<?php echo "$project_id";?> >
    <table width="100%" border="0" cellpadding="0" cellspacing="1">
        <tr>
            <td><img src="./modules/mantis/images/mantis_logo_button.gif" alt="" border="0"></td>
            <td align="left" nowrap="nowrap" width="100%"><h1><?php echo $AppUI->_( 'New Issue' );?></h1></td>
        </tr>
    </table>

    <table border="1" cellpadding="4" cellspacing="0" width="98%" class="std">

        <?PHP if ($project_id== 0){ ?>
        <tr>
            <td align="right" nowrap="nowrap"><?php echo $AppUI->_( 'Project' );?>:</td>
            <td width="60%">
                <?php
                $project = new CProject();
                $tmpprojects = $project->getAllowedProjects($AppUI->user_id);
                $projects = array();
                $projects[0] = $AppUI->_('any');
                foreach($tmpprojects as $proj) {
                    $projects[$proj['project_id']] = $proj['project_name'];
                }

                echo arraySelect( $projects, 'issue_project', 'class="text"', $project_id );
                ?>
            </td>
        </tr>
        <?php }?>
        <tr>
            <td align="right" nowrap="nowrap"><?php echo $AppUI->_( 'Summary' );?>:</td>
            <td width="80%">
                <input name="issue_summary" class="text" size="128" ></input>
            </td>
        </tr>

        <tr>
            <td align="right" nowrap="nowrap"><?php echo $AppUI->_( 'Description' );?>:</td>
            <td width="80%">
                <textarea name="issue_description" class="textarea" cols="128" rows="5" wrap="virtual"></textarea>
            </td>
        </tr>

        <tr >
            <td align="right" width="60%">
                <?php echo $AppUI->_( 'Create Task' );?>
            </td>
            <td  width="40%">
                <label><input type="radio" name='create_task' value="1" <?php echo( ON == w2PgetConfig( 'mantis_autotask') ) ? 'checked="checked" ' : ''?>/>
                <?php echo $AppUI->_( 'Yes' );?></label>
                <label><input type="radio" name='create_task' value="0" <?php echo( OFF == w2PgetConfig( 'mantis_autotask') )? 'checked="checked" ' : ''?>/>
                <?php echo $AppUI->_( 'No' );?></label>
            </td>
        </tr> 
    </table>

    <table border="0" cellspacing="0" cellpadding="3" width="98%">
        <tr>
            <td height="40" width="30%">&nbsp;</td>
            <td  height="40" width="35%" align="right">
                <table>
                <tr>
                    <td>
                    <input class="button"  type="button" value="<?php echo $AppUI->_('Cancel');?>" onclick="javascript:history.back(-1);" /> 
                    </td>
                    <td>
                        <input class="button" type="button" name="btnFuseAction" value="<?php echo $AppUI->_('save'); ?>" onClick="submit()">
                    </td>
                </tr>
                </table>
            </td>
        </tr>
    </table>
</form>
</body>
</html>